authenticate: move databroker connection to state

calebdoxsey opened the pull request:
## Summary
This PR moves the databroker connection in authenticate to the state. This way if the connection changes it will pick up the new connection automatically.

I also added a call to `ValidateOptions` every time we rebuild the state.

## Related issues
- #1254 


**Checklist**:
- [x] add related issues
- [ ] updated docs
- [x] updated unit tests
- [ ] updated UPGRADING.md
- [x] ready for review

desimone commented on authenticate/state.go:
```diff
@@ -52,6 +56,10 @@ func newAuthenticateState() *authenticateState {
 }
 
 func newAuthenticateStateFromConfig(cfg *config.Config) (*authenticateState, error) {
+	if err := ValidateOptions(cfg.Options); err != nil {
```
nit: if we make this `err` function scoped we can remove the uninitizialized err further down

desimone commented on authenticate/handlers.go:
```diff
@@ -474,10 +477,12 @@ func (a *Authenticate) getSessionFromCtx(ctx context.Context) (*sessions.State,
 		return nil
 	}
-	err := session.Delete(ctx, a.dataBrokerClient, sessionID)
+	err := session.Delete(ctx, state.dataBrokerClient, sessionID)
```
nit: I think we can just return here?

desimone commented on authenticate/handlers.go:
```diff
@@ -474,10 +477,12 @@ func (a *Authenticate) getSessionFromCtx(ctx context.Context) (*sessions.State,
-	if a.dataBrokerClient == nil {
+	state := a.state.Load()
+
+	if state.dataBrokerClient == nil {
```
If databrokerclient is nil here, do we want to error?

desimone commented on authenticate/handlers.go:
```diff
@@ -489,32 +494,34 @@ func (a *Authenticate) isAdmin(user string) bool {
 	}
 
-	pbSession, err := session.Get(r.Context(), a.dataBrokerClient, s.ID)
+	pbSession, err := session.Get(r.Context(), state.dataBrokerClient, s.ID)
```
I don't think session.Get/Set/Delete are currently handling if `dataBrokerClient` is nil?

desimone commented on authenticate/handlers.go:
```diff
@@ -144,14 +144,17 @@ func (a *Authenticate) VerifySession(next http.Handler) http.Handler {
 
-		if a.dataBrokerClient != nil {
-			_, err = session.Get(ctx, a.dataBrokerClient, sessionState.ID)
+		if state.dataBrokerClient != nil {
```
If the databroker client is nil, the request will go through? I don't think that is what we want to do here?

cuonglm commented:
> There are several instances where we do a `nil` check on the `dataBrokerClient` prior to making a series of calls to `session`/`user`/`directory`, should we move the nil check logic here to the underlying Get/Set/Delete functions instead of making it the callers responsibility to make sure the client is not nil?

I don't think we need to check that `dataBrokerClient` is nil. Technically, it can be nil, but we always initialize it in all of our code path, the only way it can be nil if we construct an instance of `Authenticate` directly, without calling update options. That will lead to more loudly and sooner panic.

calebdoxsey commented:
@cuonglm @desimone The nil check is unnecessary. It was part of the original code, but I've removed it. I believe it was added to let the tests pass, which I've now updated.
